DataStructures/AATree/AATree.cs

```diff
@@ -0,0 +1,552 @@
+        /// Gets the smallest element in the tree. (ie. the element in the left most node).
+        /// </summary>
+        /// <returns>The smallest element in the tree according to the stored comparer.</returns>
+        /// <throws>InvalidOperationException if the tree is empty.</throws>
+        public TKey GetMin()
+        {
+            if (root == null)
+            {
+                throw new InvalidOperationException("Tree is empty!");
+            }
+
+            return GetMin(root).Key;
+        }
+
+        /// <summary>
+        /// Gets all the elements in the tree in in-order order.
+        /// </summary>
+        /// <returns>Sequence of elements in in-order order.</returns>
+        public IEnumerable<TKey> GetKeysInOrder()
+        {
+            var result = new List<TKey>();
+
+            Action<AATreeNode<TKey>?>? inOrderWalk = null;
+
+            inOrderWalk = (node) =>
+            {
+                if (node == null)
+                {
+                    return;
+                }
+
+                inOrderWalk!(node.Left);
+                result.Add(node.Key);
+                inOrderWalk(node.Right);
+            };
+
+            inOrderWalk(root);
+            return result;
+        }
+
+        /// <summary>
+        /// Gets all the elements in the tree in pre-order order.
+        /// </summary>
+        /// <returns>Sequence of elements in pre-order order.</returns>
+        public IEnumerable<TKey> GetKeysPreOrder()
+        {
+            var result = new List<TKey>();
+
+            Action<AATreeNode<TKey>?>? preOrderWalk = null;
+
+            preOrderWalk = (node) =>
+            {
+                if (node == null)
+                {
+                    return;
+                }
+
+                result.Add(node.Key);
+                preOrderWalk!(node.Left);
+                preOrderWalk(node.Right);
+            };
+
+            preOrderWalk(root);
+            return result;
+        }
+
+        /// <summary>
+        /// Gets all the elements in the tree in post-order order.
+        /// </summary>
+        /// <returns>Sequence of elements in post-order order.</returns>
+        public IEnumerable<TKey> GetKeysPostOrder()
+        {
+            var result = new List<TKey>();
+
+            Action<AATreeNode<TKey>?>? postOrderWalk = null;
+
+            postOrderWalk = (node) =>
+            {
+                if (node == null)
+                {
+                    return;
+                }
+
+                postOrderWalk!(node.Left);
+                postOrderWalk(node.Right);
+                result.Add(node.Key);
+            };
+
+            postOrderWalk(root);
+            return result;
+        }
+
+        /// <summary>
+        /// Checks if the tree is a valid AA Tree.
+        /// </summary>
+        /// <returns>true if tree is a valid AA Tree, false otherwise.</returns>
+        public bool Validate()
+        {
+            return ValidateTree(root);
+        }
```

Comment:
The tree must be always valid, so no need for this method here. It can be useful in tests though.